### ISSUE TITLE
Python test failures: MPI_Comm_size() called before MPI_INIT

### ISSUE THREAD

**Issue by RJTK** [stevengj/meep#109]
I have installed meep with the Python interface from the procedure here: https://www.mail-archive.com/meep-discuss@ab-initio.mit.edu/msg05819.html

The non mpi install seems to work, and passes all the tests in make check.  However, when building with the `--with-mpi` flag passed to configure.sh, 9 of the tests fail.  Here is an excerpt from python/test-suite.log

>    meep 1.3: python/test-suite.log
> 
> \# TOTAL: 11
> \# PASS:  2
> \# SKIP:  0
> \# XFAIL: 0
> \# FAIL:  9
> \# XPASS: 0
> \# ERROR: 0
> 
> .. contents:: :depth: 2
> 
> FAIL: tests/3rd_harm_1d
> 
> *** The MPI_Comm_size() function was called before MPI_INIT was invoked.
> *** This is disallowed by the MPI standard.
> *** Your MPI job will now abort.
> [Jyan-laptop2:10071] Local abort before MPI_INIT completed successfully; not able to aggregate error messages, and not able to guarantee that all other processes were killed!

The tests bend_flux, cyl_ellipsoid, holey_wvg_bands, holey_wvg_cavity, material_dispersion, physical, ring, and source fail with the exact same message.

The tests simulation.py and geom.py are passing.

**Comment by stevengj** [comment]
What MPI implementation are you using?  The mpi4py module that we use is supposed to initialize MPI when it is loaded, but its manual says:
> Some MPI-1 implementations do require the actual command line arguments to be passed in MPI_Init(). In this case, you will need to use a rebuilt, MPI-enabled, Python interpreter executable. MPI
for Python has some support for alleviating you from this task. Check the instructions at MPI-enabled Python interpreter in the appendix.
cc @HomerReid

**Comment by RJTK** [comment]
I am using openmpi 1.8.8.  I tried reinstalling, but it resulted in the same problem.

**Comment by HomerReid** [comment]
Thanks for the bug report. I was able to reproduce your error. 

The difficulty is that the python HDF5 module (`h5py`) uses the non-MPI versions of the HDF5 libraries by default, and these conflict with the MPI versions of HDF5 used by meep when built `--with-mpi.` (This explains why some, but not all, of the python tests succeeded on your system---the ones that succeed are the ones that don't invoke HDF5 read/write routines from python). Quoting from the [`h5py` documentation](http://docs.h5py.org/en/latest/build.html):

> Be aware however that most pre-built versions lack MPI support, and that they are built against a specific version of HDF5. If you require MPI support, or newer HDF5 features, you will need to build from source.

Fortunately, the build-from-source process is relatively easy using `pip install --no-binary.` The following worked for me:

```bash
% pip uninstall h5py  
% (CC=mpicc CXX=mpic++ HDF5_MPI="ON" HDF5_DIR=/path/to/parallel/HDF5 pip install --no-binary=h5py h5py)
```

The `HDF5_DIR` assignment here is optional---use it only if the parallel HDF5 libraries are not installed in a standard location on your system.

When you get a chance, please try this out and let us know whether it works for you. We will update the installation instructions.

**Comment by stevengj** [comment]
We should try to detect this at runtime if we can, so that we can print a more comprehensible error message.

**Comment by RJTK** [comment]
I'm having trouble with h5py.  It builds and installs without issue, but upon trying to import h5py it fails with an import error "undefined symbol: ompi_mpi_comm_world".

I have ensured that mpicc and mpic++ are pointing at the right compilers, and I have tried building and install h5py directly from the newest source on github.  I believe this is some error with not using the right linking flags?  I've set `LDFLAGS=-L/usr/local/lib -L/usr/local/lib/openmpi`, `CPPFLAGS=-I/usr/local/include -I/usr/local/include/openmpi` and `LD_LIBRARY_PATH=/usr/local/lib:/usr/local/lib/openmpi`.

I've looked at `nm -D | grep *comm_world` on some of the h5py .so files, and it shows a bunch of other undefined symbols e.g.

> 00000000000acc29 T H5Pset_fapl_mpio
00000000001491b7 T H5S_mpio_space_type
0000000000170775 T H5Tcompiler_conv
                 U ompi_mpi_byte
                 U ompi_mpi_comm_null
                 U ompi_mpi_comm_self
                 U ompi_mpi_comm_world
                 U ompi_mpi_datatype_null
                 U ompi_mpi_info_null
                 U ompi_mpi_int
                 U ompi_mpi_op_bor
                 U ompi_mpi_op_sum

I also tried `nm -D /usr/local/lib/openmpi/*.so | grep "ompi_mpi*"` which turned up a bunch of other undefined symbols (excerpt):

>               U ompi_mpi_info_null
                 U ompi_mpi_int
                 U ompi_mpi_long_long_int
                 U ompi_mpi_long_long_int
                 U ompi_mpi_long_long_int
                 U ompi_mpi_comm_world
                 U ompi_mpi_unsigned_char
                 U ompi_mpi_abort
                 U ompi_mpi_byte
                 U ompi_mpi_comm_null
                 U ompi_mpi_comm_self

Does this indicate a problem with building or installing openmpi?  Where would these symbols normally get defined?  This seems odd to me, but there are no problems with `make check` when I build openmpi.

**Comment by HomerReid** [comment]
This looks to me like you might not have used the MPI C/C++ compilers when building h5py. If you prefix your build statements with (CC=mpicc CXX=mpicxx ...) you don't need to include the openmpi headers and library files in your LDFLAGS and CPPFLAGS. 

Are you sure you purged all files from previous builds before trying to build `h5py`? Could there be some parts of your installation that were built against incompatible HDF5 versions? One way to look for such things is to use `ldd` to see what shared libraries a program or library refers to. For example, 

```bash

# find the h5z shared-library file in your python installation 
% find /my/system/python/installation -iname 'h5z*so' 
./pkgs/h5py-2.7.0-np112py36_0/lib/python3.6/site-packages/h5py/h5z.cpython-36m-x86_64-linux-gnu.so

# ask what version of the HDF5 library it wants to use
% ldd `find /my/system/python/installation -iname 'h5z*so'`

./lib/python3.6/site-packages/h5py/h5z.cpython-36m-x86_64-linux-gnu.so:
	linux-vdso.so.1 =>  (0x00007fff551f7000)
-->	libhdf5.so.101 => /home/homer/codes/hdf5-parallel/lib/libhdf5.so.101 (0x00007fbd4804e000)
-->	libhdf5_hl.so.100 => /home/homer/codes/hdf5-parallel/lib/libhdf5_hl.so.100 (0x00007fbd47e2b000)**
	libpython3.6m.so.1.0 => /home/homer/codes/anaconda3/lib/libpython3.6m.so.1.0 (0x00007fbd47924000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007fbd47706000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fbd4733d000)
	libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007fbd47124000)
        ...
```

I have inserted arrows `-->` pointing out what you want to be looking at here---make sure your `h5py` installation is using the same MPI-aware HDF5 libraries that `meep` is using.
 
Meanwhile, the `pip --no-binary` command I quoted in my previous post worked for me---did you try that?

**Comment by RJTK** [comment]
Ah, I had looked at ldd, but wasn't sure what should be there.  In any case, I believe I found the problem: it is necessary to build hdf5 with `--enable-shared`, else `h5py` doesn't link with it.  It seems quite strange that h5py builds and installs without complaining, even though it never did anything with any hdf5 libraries?

meep is now building properly, and all the tests pass.

However, I believe there is still some issue, as some of the code I had that worked with the non-mpi version of meep is now failing.  I think it is a separate issue, so I will open another one.
